framework/src/meshgenerators/PatchMeshGenerator.C

```diff
@@ -0,0 +1,381 @@
+    _xoffset(getParam<Real>("x_offset")),
+    _yoffset(getParam<Real>("y_offset")),
+    _zoffset(getParam<Real>("z_offset"))
+{
+  if (_xlength <= 0.0)
+    paramError("x_length", "Must be greater than zero");
+
+  if (_ylength <= 0.0)
+    paramError("y_length", "Must be greater than zero");
+
+  if (_dim == 2 && (_elem_type == "HEX8" || _elem_type == "HEX20"))
+    paramError("elem_type", "Must be QUAD4 or QUAD8 for 2-dimensional meshes.");
+
+  if (_dim == 2 && _zlength != 0.0)
+    paramError("z_length", "Must be zero for 2-dimensional meshes");
+
+  if (_dim == 3 && !parameters.isParamSetByUser("elem_type"))
+    _elem_type = "HEX8";
+
+  if (_dim == 3 && (_elem_type == "QUAD4" || _elem_type == "QUAD8"))
+    paramError("elem_type", "Must be HEX8 or HEX20 for 3-dimensional meshes.");
+
+  // If domain dimensions are not set by user for 3-dimensions a unit cube is used
+  // as default as per MacNeal and Harder (1985)
+  if (_dim == 3)
+  {
+    if (!parameters.isParamSetByUser("x_length"))
+      _xlength = 1.0;
+    if (!parameters.isParamSetByUser("y_length"))
+      _ylength = 1.0;
+    if (!parameters.isParamSetByUser("z_length"))
+      _zlength = 1.0;
+  }
+
+  if (_dim == 3 && _zlength <= 0.0)
+    paramError("z_length", "Must be greater than zero for 3-dimensional meshes");
+}
+
+std::unique_ptr<MeshBase>
+PatchMeshGenerator::generate()
+{
+  std::unique_ptr<ReplicatedMesh> mesh = libmesh_make_unique<ReplicatedMesh>(comm(), 2);
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+
+  unsigned num_nodes = 0;
+  if (_elem_type == "QUAD4")
+    num_nodes = 8;
+  else if (_elem_type == "QUAD8")
+    num_nodes = 20;
+  else if (_elem_type == "HEX8")
+    num_nodes = 16;
+  else if (_elem_type == "HEX20")
+    num_nodes = 48;
+
+  std::vector<Node *> nodes(num_nodes);
+
+  const Real xmax = _xoffset + _xlength;
+  const Real ymax = _yoffset + _ylength;
+  const Real zmax = _zoffset + _zlength;
+
+  if (_dim == 2)
+  {
+    const std::vector<Real> x_interior_fractions{1.0 / 6.0, 3.0 / 4.0, 2.0 / 3.0, 1.0 / 3.0};
+    const std::vector<Real> y_interior_fractions{1.0 / 6.0, 1.0 / 4.0, 2.0 / 3.0, 2.0 / 3.0};
+
+    // Exterior node positions
+    std::vector<Point> node_positions{{_xoffset, _yoffset, _zoffset},
+                                      {xmax, _yoffset, _zoffset},
+                                      {xmax, ymax, _zoffset},
+                                      {_xoffset, ymax, _zoffset}};
+    // Interior node positions
+    for (unsigned i = 0; i < x_interior_fractions.size(); ++i)
+      node_positions.push_back({_xoffset + x_interior_fractions[i] * _xlength,
+                                _yoffset + y_interior_fractions[i] * _ylength,
+                                _zoffset});
+
+    if (_elem_type == "QUAD8")
+    {
+      // Exterior midside node positions
+      for (unsigned i = 0; i < 4; ++i)
+        node_positions.push_back({0.5 * (node_positions[i](0) + node_positions[(i + 1) % 4](0)),
```

Comment:
Did you know that all of the arithmetic operators are defined on Point? No need to do the same operation on every component. All of these should be simplified to one-liners!